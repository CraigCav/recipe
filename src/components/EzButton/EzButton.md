---
name: EzButton
title: Button
path: '/components/ez-button'
---

Buttons represent actions on a page that can be triggered with one click. Buttons can be used in forms, or in other locations in a page to communicate that an action is available.

---

## Best Practices

Buttons should:

- Be clearly labeled, or provide an accessible name (icon buttons)
- Be positioned in consistent locations
- Use spacing to separate them from other interactive content
- Not wrap text. For maximum legibility, a label should remain on a single line

Buttons should not:

- Be used be used in place of a more semantic element. For example, favor anchors for navigation.

---

## Content guidelines

Buttons should clearly communicate the action that will occur when the user interacts with them. Try leading with actionable verbs when labeling Buttons. Follow the Verb+Noun format on buttons except in the case of common actions like Save, Edit, Close, or Cancel.

## Examples

### Primary Button

Primary buttons are used to highlight the most important actions a user can take. There should not be more than one primary button in a section of a page.

```jsx
<EzButton use="primary">Create Order</EzButton>
```

### Secondary Button

Secondary buttons are used for action on a page that are important, but aren't the primary action.

```jsx
<EzButton use="secondary">Edit</EzButton>
```

### Tertiary Button

Tertiary buttons are used when a section or page has too many actions of varying importance. A page should always have one primary action, and may have one or two secondary actions. Any actions that aren’t as important as these primary and secondary actions should use a tertiary button style. For example, on an order details page, if we’re asking the user to accept the order (primary), and reject or request ezDispatch (secondary actions), then any other actions aren’t as important and should use a tertiary button.

Tertiary buttons should also be used when space is a constraint or when content is repeated many times on a page. For example, an action in a table row should use a tertiary button since space is limited and having the outline around the button text in multiple table rows would distract the user from the action we want them to take.

```jsx
<EzButton use="tertiary">Save as PDF</EzButton>
```

### Destructive Button

When an action will delete data, or discard the current page state, a destructive button should be used. Destructive buttons should trigger a confirmation dialog before the action is completed. Destructive buttons should typically be secondary buttons, but are promoted to primary buttons on confirmation dialogs.

```jsx
<EzLayout layout="basic">
  <EzButton use="secondary" destructive>
    Delete Draft
  </EzButton>
  <EzButton use="primary" destructive>
    Confirm Deletion
  </EzButton>
  <EzButton use="tertiary" destructive>
    Delete Item
  </EzButton>
</EzLayout>
```

### Disabled State

Used to indicate to the user that an action is not currently available. The surrounding interface should indicate the necessary prerequisite action to enable the action. Disabled buttons do not respond to user input (and therefore will not trigger onClick behavior).

```jsx
<div>
  <EzButton use="primary" disabled>
    Save Changes
  </EzButton>
</div>
```

### Loading State

Used to indicate to the user that an action is currently being processed. The button remains disabled and no further actions will be triggered until loading has completed.

```jsx
<div>
  <EzButton use="primary" loading>
    Save Changes
  </EzButton>
</div>
```
