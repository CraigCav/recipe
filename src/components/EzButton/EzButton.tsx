import React, {createElement, forwardRef} from 'react';
import StyledButton, {DisabledButtonWrapper} from './EzButton.styles';
import EzTooltip from '../EzTooltip';

type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & (RegularProps | TertiaryProps);

type SharedButtonProps = {
  destructive?: boolean;
  disabled?: boolean;
  disabledMessage?: string;
  onClick?: React.MouseEventHandler;
};

type RegularProps = {
  loading?: boolean;
  use: 'primary' | 'secondary';
} & SharedButtonProps;

type TertiaryProps = {
  use: 'tertiary';
} & SharedButtonProps;

/**
 * Buttons represent actions on a page that can be triggered with one click.
 * Buttons can be used in forms, or in other locations in a page to communicate that an action is available.
 */
const EzButton = forwardRef<HTMLElement, ButtonProps>((props, ref) => {
  const buttonElement = createElement(StyledButton, buildProps({...props, innerRef: ref}));

  if (props.disabled && props.disabledMessage) {
    return (
      <EzTooltip message={props.disabledMessage}>
        <DisabledButtonWrapper>{buttonElement}</DisabledButtonWrapper>
      </EzTooltip>
    );
  }

  return buttonElement;
});

const buildProps = (props: ButtonProps & {innerRef: React.Ref<HTMLElement>}) => {
  if (props.use === 'tertiary') return props;

  return {...props, disabled: props.disabled || props.loading};
};

/**
 * defaultProps
 * @property {string} use - Buttons are considered secondary by default.
 */
EzButton.defaultProps = {
  use: 'secondary',
};

EzButton.displayName = 'EzButton';

/**
 * @component
 */
export default EzButton;
