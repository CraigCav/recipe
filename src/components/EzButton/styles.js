import {css} from 'react-emotion';
import {shade} from 'polished';

export const base = ({theme}) => css`
  border: none;
  border-radius: ${theme.borderRadius[1]};
  cursor: pointer;
  display: inline-block;
  outline: 0;
  font-family: Lato, 'Helvetica Neue', Arial, Helvetica, sans-serif;
  line-height: 1em;
  min-height: 1em;
  margin-right: ${theme.spacing[0]};
  padding: ${theme.spacing[2]} ${theme.spacing[4]};
  text-decoration: none;
  user-select: none;
  -webkit-tap-highlight-color: transparent;
`;

export const disabled = ({disabled}) =>
  disabled &&
  css`
    cursor: default;
    opacity: 0.45;
    pointer-events: none;
  `;

const elementStates = (style, color) => ({
  [style]: color,
  ['&:hover:enabled']: {[style]: shade(0.9, color)},
  ['&:focus:enabled']: {[style]: shade(0.8, color)},
  ['&:active:enabled']: {[style]: shade(0.7, color)},
});

export const primary = ({theme: {colors}, destructive}) => {
  const {main} = destructive ? colors.destructive : colors.primary;
  return css`
    color: ${colors.primary.contrastText};
    ${elementStates('background-color', main)};
  `;
};

export const secondary = ({theme: {colors}, destructive}) => {
  const {main} = destructive ? colors.destructive : colors.primary;
  return css`
    background-color: transparent;
    border: solid 1px;
    ${elementStates('border-color', main)};
    ${elementStates('color', main)};
  `;
};
