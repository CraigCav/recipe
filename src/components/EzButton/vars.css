:root {
  --recipe-button-font-size: var(--recipe-alias-font-size-default);
  --recipe-button-font-family: var(--recipe-global-font-family-base);
  --recipe-button-font-weight: var(--recipe-global-font-weight-bold);
  --recipe-button-leading: var(--recipe-global-leading-tight);
  --recipe-button-border-radius: var(--recipe-alias-border-radius-regular);
  --recipe-button-border-size: var(--recipe-alias-border-size-thin);
  --recipe-button-padding-x: var(--recipe-global-static-size-200);
  --recipe-button-padding-y: var(--recipe-global-static-size-75);

  /* primary */
  --recipe-button-primary-background-color: var(--recipe-global-color-static-blue-600);
  --recipe-button-primary-background-color-hover: var(--recipe-global-color-static-blue-800);
  --recipe-button-primary-background-color-down: var(--recipe-global-color-static-blue-900);
  --recipe-button-primary-background-color-focus: var(--recipe-global-color-static-blue-650);

  --recipe-button-primary-text-color: var(--recipe-global-color-static-white);
  --recipe-button-primary-text-color-hover: var(--recipe-global-color-static-white);
  --recipe-button-primary-text-color-down: var(--recipe-global-color-static-white);
  
  --recipe-button-primary-border-color: var(--recipe-global-color-static-blue-600);
  --recipe-button-primary-border-color-hover: var(--recipe-global-color-static-blue-800);
  --recipe-button-primary-border-color-down: var(--recipe-global-color-static-blue-900);
  --recipe-button-primary-border-color-focus: var(--recipe-global-color-static-blue-950);

  /* primary--destructive */
  --recipe-button-primary-destructive-background-color: var(--recipe-global-color-static-red-600);
  --recipe-button-primary-destructive-background-color-hover: var(--recipe-global-color-static-red-800);
  --recipe-button-primary-destructive-background-color-down: var(--recipe-global-color-static-red-900);
  --recipe-button-primary-destructive-background-color-focus: var(--recipe-global-color-static-red-600);

  --recipe-button-primary-destructive-text-color: var(--recipe-global-color-static-white);
  --recipe-button-primary-destructive-text-color-hover: var(--recipe-global-color-static-white);
  --recipe-button-primary-destructive-text-color-down: var(--recipe-global-color-static-white);
  
  --recipe-button-primary-destructive-border-color: var(--recipe-global-color-static-red-600);
  --recipe-button-primary-destructive-border-color-hover: var(--recipe-global-color-static-red-800);
  --recipe-button-primary-destructive-border-color-down: var(--recipe-global-color-static-red-900);
  --recipe-button-primary-destructive-border-color-focus: var(--recipe-global-color-static-red-950);

  /* secondary */
  --recipe-button-secondary-background-color: var(--recipe-global-color-static-white);
  --recipe-button-secondary-background-color-hover: var(--recipe-global-color-static-white);
  --recipe-button-secondary-background-color-down: var(--recipe-global-color-static-white);
  --recipe-button-secondary-background-color-focus: var(--recipe-global-color-static-white);

  --recipe-button-secondary-text-color: var(--recipe-global-color-static-blue-600);
  --recipe-button-secondary-text-color-hover: var(--recipe-global-color-static-blue-800);
  --recipe-button-secondary-text-color-down: var(--recipe-global-color-static-blue-900);
  
  --recipe-button-secondary-border-color: var(--recipe-global-color-static-gray-400);
  --recipe-button-secondary-border-color-hover: var(--recipe-global-color-static-gray-500);
  --recipe-button-secondary-border-color-down: var(--recipe-global-color-static-gray-400);
  --recipe-button-secondary-border-color-focus: var(--recipe-global-color-blue-gray-600);

  /* secondary--destructive */
  --recipe-button-secondary-destructive-background-color: var(--recipe-global-color-static-white);
  --recipe-button-secondary-destructive-background-color-hover: var(--recipe-global-color-static-white);
  --recipe-button-secondary-destructive-background-color-down: var(--recipe-global-color-static-white);
  --recipe-button-secondary-destructive-background-color-focus: var(--recipe-global-color-static-white);

  --recipe-button-secondary-destructive-text-color: var(--recipe-global-color-static-red-600);
  --recipe-button-secondary-destructive-text-color-hover: var(--recipe-global-color-static-red-800);
  --recipe-button-secondary-destructive-text-color-down: var(--recipe-global-color-static-red-900);
  
  --recipe-button-secondary-destructive-border-color: var(--recipe-global-color-static-gray-400);
  --recipe-button-secondary-destructive-border-color-hover: var(--recipe-global-color-static-gray-500);
  --recipe-button-secondary-destructive-border-color-down: var(--recipe-global-color-static-gray-400);
  --recipe-button-secondary-destructive-border-color-focus: var(--recipe-global-color-static-blue-600);

  /* tertiary */
  --recipe-button-tertiary-text-color: var(--recipe-global-color-static-blue-600);
  --recipe-button-tertiary-text-color-hover: var(--recipe-global-color-static-blue-800);
  --recipe-button-tertiary-text-color-down: var(--recipe-global-color-static-blue-900);

  /* tertiary--destructive */
  --recipe-button-tertiary-destructive-text-color: var(--recipe-global-color-static-red-600);
  --recipe-button-tertiary-destructive-text-color-hover: var(--recipe-global-color-static-red-800);
  --recipe-button-tertiary-destructive-text-color-down: var(--recipe-global-color-static-red-900);
}