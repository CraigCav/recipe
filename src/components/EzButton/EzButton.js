import {createElement} from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';
import variant from 'styled-component-variant';
import {base, primary, secondary, tertiary, disabled, loading, outline} from './EzButton.styles';
import {standard} from '../../themes';

const use = variant('use', {
  primary,
  secondary,
  tertiary,
});

const baseStyles = [base, use, disabled, loading, outline];

/**
 * Buttons represent actions on a page that can be triggered with one click.
 * Buttons can be used in forms, or in other locations in a page to communicate that an action is available.
 */
const EzButton = props => {
  const StyledButton = styled('button')(...baseStyles);
  return createElement(StyledButton, {
    ...props,
    disabled: props.disabled || props.loading,
  });
};

EzButton.propTypes = {
  /**
   * If `true`, indicates that the button is used for actions that delete, disconnect or perform an irreversible action.
   */
  destructive: PropTypes.bool,
  /**
   * If `true`, indicates that the button is currently unable to be interacted with.
   */
  disabled: PropTypes.bool,
  /**
   * If `true`, indicates that the button action is being processed and that the button is currently unavailable to be interacted with.
   * Cannot be used in combination with tertiary button types.
   */
  loading(props) {
    if (props.use === 'tertiary' && props.loading)
      return new Error("Don't use loading in combination with tertiary");

    if (!['undefined', 'boolean'].includes(typeof props.loading))
      return new Error('loading must be a boolean if provided');

    return undefined;
  },
  /**
   * Determines the emphasis of the button.
   */
  use: PropTypes.oneOf(['primary', 'secondary', 'tertiary']).isRequired,
};

/**
 * defaultProps
 * @property {string} use - Buttons are considered secondary by default.
 * @property {object} theme - used the standard theme by default.
 */
EzButton.defaultProps = {
  use: 'secondary',
  theme: standard,
};

EzButton.displayName = 'EzButton';

/**
 * @component
 */
export default EzButton;
