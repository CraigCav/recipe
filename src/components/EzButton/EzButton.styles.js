import styled, {css} from 'react-emotion';
import {shade} from 'polished';
import variant from 'styled-component-variant';
import {keyframes} from '../../styles';

const base = ({theme}) => css`
  cursor: pointer;
  display: inline-block;
  outline: 0;
  font-family: Lato, 'Helvetica Neue', Arial, Helvetica, sans-serif;
  font-weight: ${theme.fontWeights.bold};
  line-height: 1em;
  min-height: 1em;
  margin-right: ${theme.spacing.xs2};
  text-decoration: none;
  user-select: none;
  white-space: nowrap;
  -webkit-tap-highlight-color: transparent;
`;

const disabled = ({disabled: isDisabled}) =>
  isDisabled &&
  css`
    cursor: default;
    opacity: 0.45;
    pointer-events: none;
  `;

const elementStates = (style, color, focusColor) => ({
  [style]: color,
  '&:hover:enabled': {[style]: shade(0.85, color)},
  '&:focus:enabled': focusColor && {[style]: focusColor},
  '&:active:enabled': {[style]: shade(0.75, color)},
});

const outlineStyles = ({theme}) => css`
  border-radius: ${theme.borderRadius[1]};
  border-style: solid;
  border-width: ${theme.borderWidth[0]};
  box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.12);
  padding: ${theme.spacing.xs} ${theme.spacing.md};
`;

const outline = variant('use', {
  primary: outlineStyles,
  secondary: outlineStyles,
});

const primary = ({theme: {colors}, destructive}) => {
  const {main} = destructive ? colors.destructive : colors.primary;
  return css`
    color: ${colors.primary.contrastText};
    border-color: ${main};
    ${elementStates('border-color', main, shade(0.75, main))};
    ${elementStates('background-color', main)};
  `;
};

const secondary = ({theme: {colors}, destructive}) => {
  const {main} = destructive ? colors.destructive : colors.primary;
  return css`
    background-color: white;
    border: solid 1px;
    ${elementStates('border-color', colors.grays[400], colors.blues[600])};
    ${elementStates('color', main)};
  `;
};

const tertiary = ({theme: {colors}, destructive}) => {
  const {main} = destructive ? colors.destructive : colors.primary;
  return css`
    background: none;
    border: none;
    ${elementStates('color', main)};
  `;
};

const spinnerSize = 1.28571429;
const spinnerOptions = {
  size: `${spinnerSize}rem`,
  margin: `${spinnerSize / 2}rem`,
  thickness: '0.2rem',
  radius: '10rem',
};

const spinner = ({margin, size, radius, thickness, color}) => css`
  position: relative;
  color: transparent;

  &::before {
    position: absolute;
    content: '';
    top: 50%;
    left: 50%;
    margin: -${margin} 0 0 -${margin};
    width: ${size};
    height: ${size};
    border-radius: ${radius};
    border: ${thickness} solid rgba(0, 0, 0, 0.15);
  }

  &::after {
    position: absolute;
    content: '';
    top: 50%;
    left: 50%;
    margin: -${margin} 0 0 -${margin};
    width: ${size};
    height: ${size};
    animation: ${keyframes.spin()} 0.6s linear;
    animation-iteration-count: infinite;
    border-radius: ${radius};
    border-color: ${color} transparent transparent;
    border-style: solid;
    border-width: ${thickness};
    box-shadow: 0 0 0 1px transparent;
  }
`;

const loading = ({loading: isLoading, use, theme: {colors}}) => {
  const color = variant('use', {
    primary: colors.white,
    secondary: colors.grays[700],
  })({use});
  return isLoading && spinner({...spinnerOptions, color});
};

const use = variant('use', {
  primary,
  secondary,
  tertiary,
});

const baseStyles = [base, use, disabled, loading, outline];

export default styled('button')(...baseStyles);
