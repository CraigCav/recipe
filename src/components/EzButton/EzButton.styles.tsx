import {css} from 'react-emotion';
import variant from 'styled-component-variant';
import styled from '../../themes/styled';
import {keyframes, darken, pseudoClasses} from '../../styles';

const base = ({theme}) => css`
  cursor: pointer;
  display: inline-block;
  outline: 0;
  font-family: Lato, 'Helvetica Neue', Arial, Helvetica, sans-serif;
  font-weight: ${theme.fontWeights.bold};
  line-height: 1em;
  min-height: 1em;
  text-decoration: none;
  user-select: none;
  white-space: nowrap;
  -webkit-tap-highlight-color: transparent;
`;

const disabled = ({disabled: isDisabled}) =>
  isDisabled &&
  css`
    cursor: default;
    opacity: 0.45;
    pointer-events: none;
  `;

const outlineStyles = ({theme}) => css`
  border-radius: ${theme.borderRadius[1]};
  border-style: solid;
  border-width: ${theme.borderWidth[0]};
  box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.12);
  padding: ${theme.spacing.xs} ${theme.spacing.md};
`;

const outline = variant('use', {
  primary: outlineStyles,
  secondary: outlineStyles,
});

const primary = ({theme: {colors}, destructive}) => {
  const {main} = destructive ? colors.destructive : colors.primary;
  return css`
    color: ${colors.primary.contrastText};
    border-color: ${main};
    ${pseudoClasses('border-color', {color: main, focusColor: darken(main, 0.25)})};
    ${pseudoClasses('background-color', {color: main})};
  `;
};

const secondary = ({theme: {colors}, destructive}) => {
  const {main} = destructive ? colors.destructive : colors.primary;
  return css`
    background-color: white;
    border: solid 1px;
    ${pseudoClasses('border-color', {color: colors.grays[400], focusColor: colors.blues[600]})};
    ${pseudoClasses('color', {color: main})};
  `;
};

const tertiary = ({theme: {colors}, destructive}) => {
  const {main} = destructive ? colors.destructive : colors.primary;
  return css`
    background: none;
    border: none;
    padding: 0;
    ${pseudoClasses('color', {color: main})};

    &:focus {
      outline: auto;
    }
  `;
};

const spinnerSize = 1.28571429;
const spinnerOptions = {
  size: `${spinnerSize}rem`,
  margin: `${spinnerSize / 2}rem`,
  thickness: '0.2rem',
  radius: '10rem',
};

const spinner = ({margin, size, radius, thickness, color}) => css`
  position: relative;
  color: transparent;

  &::before {
    position: absolute;
    content: '';
    top: 50%;
    left: 50%;
    margin: -${margin} 0 0 -${margin};
    width: ${size};
    height: ${size};
    border-radius: ${radius};
    border: ${thickness} solid rgba(0, 0, 0, 0.15);
  }

  &::after {
    position: absolute;
    content: '';
    top: 50%;
    left: 50%;
    margin: -${margin} 0 0 -${margin};
    width: ${size};
    height: ${size};
    animation: ${keyframes.spin()} 0.6s linear;
    animation-iteration-count: infinite;
    border-radius: ${radius};
    border-color: ${color} transparent transparent;
    border-style: solid;
    border-width: ${thickness};
    box-shadow: 0 0 0 1px transparent;
  }
`;

const loading = ({loading: isLoading, use, theme: {colors}}) => {
  const color = variant('use', {
    primary: colors.white,
    secondary: colors.grays[700],
  })({use});
  return isLoading && spinner({...spinnerOptions, color});
};

const use = variant('use', {
  primary,
  secondary,
  tertiary,
});

const baseStyles = [base, use, disabled, loading, outline];

export default styled('button')(...baseStyles);
