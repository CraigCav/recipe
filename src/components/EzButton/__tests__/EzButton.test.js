import React from 'react';
import EzButton from '../EzButton';

describe('EzButton', () => {
  let actual;

  beforeEach(() => (actual = shallow(<EzButton use="primary">Click Me</EzButton>)));

  /**
   * Style tests.
   */
  it('should render with primary styles', () => {
    expect(actual).toMatchSnapshot();
  });

  it('should render with secondary styles', () => {
    actual.setProps({use: 'secondary'});
    expect(actual).toMatchSnapshot();
  });

  it('should render with tertiary styles', () => {
    actual.setProps({use: 'tertiary'});
    expect(actual).toMatchSnapshot();
  })

  it('should render with disabled styles', () => {
    actual.setProps({disabled: true});
    expect(actual).toMatchSnapshot();
  });

  it('should render primary button with destructive styles', () => {
    actual.setProps({destructive: true});
    expect(actual).toMatchSnapshot();
  });

  it('should render secondary button with destructive styles', () => {
    actual.setProps({destructive: true, use: 'secondary'});
    expect(actual).toMatchSnapshot();
  });

  it('should render tertiary button with destructive styles', () => {
    actual.setProps({destructive: true, use: 'tertiary'});
    expect(actual).toMatchSnapshot();
  });

  it('should render primary button with loading styles', () => {
    actual.setProps({loading: true, use: 'primary'});
    expect(actual).toMatchSnapshot();
  });

  it('should render secondary button with loading styles', () => {
    actual.setProps({loading: true, use: 'secondary'});
    expect(actual).toMatchSnapshot();
  });

  /**
   * Logic tests.
   */
  it('renders a button element by default', () => {
    expect(actual.name()).toEqual('button');
  });

  describe('disabled', () => {
    it('is not disabled by default', () => {
      expect(actual.prop('disabled')).toBeUndefined();
    });

    it('is applied to the button element when defined', () => {
      actual.setProps({disabled: true});
      expect(actual.prop('disabled')).toBe(true);
    });

    it('is does not trigger clicks', () => {
      const spy = jest.fn();

      const component = mount(
        <EzButton use="primary" onClick={spy} disabled>
          Click Me
        </EzButton>
      );

      component.simulate('click');

      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('loading', () => {
    it('is not loading by default', () => {
      expect(actual.prop('loading')).toBeUndefined();
    });

    it('is applies the disabled attribute to the button element', () => {
      actual.setProps({disabled: true});
      expect(actual.prop('disabled')).toBe(true);
    });

    it('is does not trigger clicks', () => {
      const spy = jest.fn();

      const component = mount(
        <EzButton use="primary" onClick={spy} loading>
          Click Me
        </EzButton>
      );

      component.simulate('click');

      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('valid props', () => {
    it('renders valid props for html elements', () => {
      actual.setProps({'data-test': 'my-test-selector'});
      expect(actual.prop('data-test')).toEqual('my-test-selector');
    });

    it('does NOT render invalid props for html elements', () => {
      expect(actual.prop('primary')).toBeUndefined();
    });
  });

  /**
   * Accessibility tests.
   */
  it('should meet accessibility guidelines for buttons', async () => {
    const wrapper = renderToHtml(<EzButton use="primary">Click Me</EzButton>);
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
