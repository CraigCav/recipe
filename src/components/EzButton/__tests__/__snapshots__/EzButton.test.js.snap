// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EzButton should render primary button with destructive styles 1`] = `
.recipe-0 {
  border: 1px solid;
  border-radius: 4px;
  box-shadow: 0 1px 1px 0 rgba(0,0,0,0.12);
  cursor: pointer;
  display: inline-block;
  outline: 0;
  font-family: Lato,'Helvetica Neue',Arial,Helvetica,sans-serif;
  line-height: 1em;
  min-height: 1em;
  margin-right: 0.25rem;
  padding: 0.75rem 1.5rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-tap-highlight-color: transparent;
  color: #FFFFFF;
  border-color: #db2828;
  border-color: #db2828;
  background-color: #db2828;
}

.recipe-0:hover:enabled {
  border-color: #ba2222;
}

.recipe-0:focus:enabled {
  border-color: #a41e1e;
}

.recipe-0:active:enabled {
  border-color: #a41e1e;
}

.recipe-0:hover:enabled {
  background-color: #ba2222;
}

.recipe-0:active:enabled {
  background-color: #a41e1e;
}

<button
  className="recipe-0 recipe-1"
>
  Click Me
</button>
`;

exports[`EzButton should render secondary button with destructive styles 1`] = `
.recipe-0 {
  border: 1px solid;
  border-radius: 4px;
  box-shadow: 0 1px 1px 0 rgba(0,0,0,0.12);
  cursor: pointer;
  display: inline-block;
  outline: 0;
  font-family: Lato,'Helvetica Neue',Arial,Helvetica,sans-serif;
  line-height: 1em;
  min-height: 1em;
  margin-right: 0.25rem;
  padding: 0.75rem 1.5rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-tap-highlight-color: transparent;
  background-color: white;
  border: solid 1px;
  border-color: #CED4D9;
  color: #db2828;
}

.recipe-0:hover:enabled {
  border-color: #afb4b8;
}

.recipe-0:focus:enabled {
  border-color: #3E90D6;
}

.recipe-0:active:enabled {
  border-color: #9a9fa2;
}

.recipe-0:hover:enabled {
  color: #ba2222;
}

.recipe-0:active:enabled {
  color: #a41e1e;
}

<button
  className="recipe-0 recipe-1"
>
  Click Me
</button>
`;

exports[`EzButton should render with disabled styles 1`] = `
.recipe-0 {
  border: 1px solid;
  border-radius: 4px;
  box-shadow: 0 1px 1px 0 rgba(0,0,0,0.12);
  cursor: pointer;
  display: inline-block;
  outline: 0;
  font-family: Lato,'Helvetica Neue',Arial,Helvetica,sans-serif;
  line-height: 1em;
  min-height: 1em;
  margin-right: 0.25rem;
  padding: 0.75rem 1.5rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-tap-highlight-color: transparent;
  color: #FFFFFF;
  border-color: #2484d7;
  border-color: #2484d7;
  background-color: #2484d7;
  cursor: default;
  opacity: 0.45;
  pointer-events: none;
}

.recipe-0:hover:enabled {
  border-color: #1e70b6;
}

.recipe-0:focus:enabled {
  border-color: #1b63a1;
}

.recipe-0:active:enabled {
  border-color: #1b63a1;
}

.recipe-0:hover:enabled {
  background-color: #1e70b6;
}

.recipe-0:active:enabled {
  background-color: #1b63a1;
}

<button
  className="recipe-0 recipe-1"
  disabled={true}
>
  Click Me
</button>
`;

exports[`EzButton should render with primary styles 1`] = `
.recipe-0 {
  border: 1px solid;
  border-radius: 4px;
  box-shadow: 0 1px 1px 0 rgba(0,0,0,0.12);
  cursor: pointer;
  display: inline-block;
  outline: 0;
  font-family: Lato,'Helvetica Neue',Arial,Helvetica,sans-serif;
  line-height: 1em;
  min-height: 1em;
  margin-right: 0.25rem;
  padding: 0.75rem 1.5rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-tap-highlight-color: transparent;
  color: #FFFFFF;
  border-color: #2484d7;
  border-color: #2484d7;
  background-color: #2484d7;
}

.recipe-0:hover:enabled {
  border-color: #1e70b6;
}

.recipe-0:focus:enabled {
  border-color: #1b63a1;
}

.recipe-0:active:enabled {
  border-color: #1b63a1;
}

.recipe-0:hover:enabled {
  background-color: #1e70b6;
}

.recipe-0:active:enabled {
  background-color: #1b63a1;
}

<button
  className="recipe-0 recipe-1"
>
  Click Me
</button>
`;

exports[`EzButton should render with secondary styles 1`] = `
.recipe-0 {
  border: 1px solid;
  border-radius: 4px;
  box-shadow: 0 1px 1px 0 rgba(0,0,0,0.12);
  cursor: pointer;
  display: inline-block;
  outline: 0;
  font-family: Lato,'Helvetica Neue',Arial,Helvetica,sans-serif;
  line-height: 1em;
  min-height: 1em;
  margin-right: 0.25rem;
  padding: 0.75rem 1.5rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-tap-highlight-color: transparent;
  background-color: white;
  border: solid 1px;
  border-color: #CED4D9;
  color: #2484d7;
}

.recipe-0:hover:enabled {
  border-color: #afb4b8;
}

.recipe-0:focus:enabled {
  border-color: #3E90D6;
}

.recipe-0:active:enabled {
  border-color: #9a9fa2;
}

.recipe-0:hover:enabled {
  color: #1e70b6;
}

.recipe-0:active:enabled {
  color: #1b63a1;
}

<button
  className="recipe-0 recipe-1"
>
  Click Me
</button>
`;
